src/uuid/mod.rs

```diff
@@ -0,0 +1,461 @@
+
+pub(crate) const UUID_NEWTYPE_NAME: &str = "BsonUuid";
+
+/// A struct modeling a BSON UUID value (i.e. a Binary value with subtype 4).
+///
+/// This type should be used instead of [`uuid::Uuid`](https://docs.rs/uuid/latest/uuid/struct.Uuid.html)
+/// when serializing to or deserializing from BSON, since
+/// [`uuid::Uuid`](https://docs.rs/uuid/latest/uuid/struct.Uuid.html)'s `serde` implementation doesn't
+/// produce or parse BSON UUIDs.
+///
+/// To enable interop with the `Uuid` type from the `uuid` crate, enable the `uuid-0_8` feature
+/// flag.
+///
+/// For more information on the usage of this type, see the [`uuid`] module-level documentation.
+#[derive(Clone, Copy, PartialEq, Hash)]
+pub struct Uuid {
+    uuid: uuid::Uuid,
+}
+
+impl Uuid {
+    /// Creates a random UUID.
+    ///
+    /// This uses the operating system's RNG as the source of random numbers. If you'd like to use a
+    /// custom generator, generate random bytes and pass them to [`Uuid::from_bytes`] instead.
+    pub fn new() -> Self {
+        Self {
+            uuid: uuid::Uuid::new_v4(),
+        }
+    }
+
+    /// Creates a [`Uuid`] using the supplied big-endian bytes.
+    pub const fn from_bytes(bytes: [u8; 16]) -> Self {
+        Self::from_external_uuid(uuid::Uuid::from_bytes(bytes))
+    }
+
+    /// Creates a [`Uuid`] from the provided hex string.
+    pub fn parse_str(input: impl AsRef<str>) -> Result<Self> {
+        let uuid = uuid::Uuid::parse_str(input.as_ref()).map_err(|e| Error::InvalidUuidString {
+            message: e.to_string(),
+        })?;
+        Ok(Self::from_external_uuid(uuid))
+    }
+
+    pub(crate) const fn from_external_uuid(uuid: uuid::Uuid) -> Self {
+        Self { uuid }
+    }
+
+    /// Returns an array of 16 bytes containing the [`Uuid`]'s data.
+    pub const fn bytes(self) -> [u8; 16] {
+        *self.uuid.as_bytes()
+    }
+}
+
+impl Default for Uuid {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(feature = "uuid-0_8")]
+#[cfg_attr(docsrs, doc(cfg(feature = "uuid-0_8")))]
+impl Uuid {
+    /// Convert this [`Uuid`] to a [`uuid::Uuid`] from the [`uuid`](https://docs.rs/uuid/latest) crate.
+    pub fn from_uuid_0_8(uuid: uuid::Uuid) -> Self {
+        Self::from_external_uuid(uuid)
+    }
+
+    /// Convert this [`Uuid`] to a [`uuid::Uuid`] from the [`uuid`](https://docs.rs/uuid/latest) crate.
+    pub fn to_uuid_0_8(self) -> uuid::Uuid {
+        self.uuid
+    }
+}
+
+impl Serialize for Uuid {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_newtype_struct(UUID_NEWTYPE_NAME, &self.uuid)
+    }
+}
+
+impl<'de> Deserialize<'de> for Uuid {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        match deserializer.deserialize_newtype_struct(UUID_NEWTYPE_NAME, BsonVisitor)? {
```

Comment:
deserialization works by signaling to the deserializer, again through the newtype name, that we're looking for a UUID. The BSON deserializers ensure that a binary subtype 4 is present in the data and then pass it on. Non-BSON deserializers will just attempt to deserialize a `Bson` normally, and from there we deserialize a `uuid::Uuid` manually.